cmd/gitlab.go

```diff
@@ -386,86 +362,145 @@ func (c *Client) pollPipelineJobs(p Project, pipelineID int, topics string, ref
+			}
+		}
 
-				if p.ShouldFetchPipelineJobMetrics(cfg) {
-					if err := c.pollPipelineJobs(p, lastPipeline.ID, topics, ref); err != nil {
-						log.Errorf("Could not poll jobs for pipeline %d: %s", lastPipeline.ID, err.Error())
-					}
-				}
+		timeSinceLastRun.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(time.Since(*lastPipeline.CreatedAt).Round(time.Second).Seconds())
+	}
 
+	return nil
+}
+
+func (c *Client) pollProjects(until <-chan bool) {
+	go func() {
+		pollWildcardsEvery := time.NewTicker(time.Duration(cfg.ProjectsPollingIntervalSeconds) * time.Second)
+		pollRefsEvery := time.NewTicker(time.Duration(cfg.RefsPollingIntervalSeconds) * time.Second)
+		stopWorkers := make(chan struct{})
+		defer close(stopWorkers)
+		// first execution
+		c.discoverWildcards()
+		c.pollWithWorkersUntil(stopWorkers)
+		for {
+			select {
+			case <-until:
+				log.Info("stopping projects polling...")
+				return
+			case <-pollRefsEvery.C:
+				c.pollWithWorkersUntil(stopWorkers)
+			case <-pollWildcardsEvery.C:
+				c.discoverWildcards()
 			}
 		}
+	}()
+}
 
-		if lastPipeline != nil {
-			timeSinceLastRun.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(float64(time.Since(*lastPipeline.CreatedAt).Round(time.Second).Seconds()))
-			b.Reset()
-		}
+func (c *Client) discoverWildcards() {
+	log.Infof("%d wildcard(s) configured for polling", len(cfg.Wildcards))
+	if err := c.pollProjectsFromWildcards(); err != nil {
+		log.Errorf("%s", err.Error())
+	}
+}
 
-		time.Sleep(b.Duration())
+func (c *Client) pollWithWorkersUntil(stopWorkers <-chan struct{}) {
+	log.Infof("%d project(s) configured for polling", len(cfg.Projects))
+	pollErrors := c.pollProjectsWithMaxWorkers(stopWorkers, cfg.Projects...)
+	for _, err := range pollErrors {
+		log.Errorf("%v", err)
 	}
 }
 
-func (c *Client) pollProjects() {
-	log.Infof("%d project(s) and %d wildcard(s) configured", len(cfg.Projects), len(cfg.Wildcards))
-	for _, p := range cfg.Projects {
-		go c.pollProject(p)
+func (c *Client) pollProjectsFromWildcards() error {
+	for _, w := range cfg.Wildcards {
+		foundProjects, err := c.listProjects(&w)
+		if err != nil {
+			return fmt.Errorf("could not list projects for wildcard %v: %v", w, err)
+		}
+		for _, p := range foundProjects {
+			if !projectExists(p) {
+				log.Infof("Found new project: %s", p.Name)
+				//go c.pollProject(p) we don't need to poll projects because they will be picked up at next iteration
+				cfg.Projects = append(cfg.Projects, p)
+			}
+		}
 	}
+	return nil
+}
 
-	for {
-		c.pollProjectsFromWildcards()
-		time.Sleep(time.Duration(cfg.ProjectsPollingIntervalSeconds) * time.Second)
+func (c *Client) pollProjectsWithMaxWorkers(until <-chan struct{}, projects ...Project) []error {
+	var errs []error
+	errorStream := make(chan error)
+	defer close(errorStream)
+	projectsToPoll := make(chan Project, len(projects))
+	// spawn GOMAXPROCS workers to process project polling
+	for w := 0; w < runtime.GOMAXPROCS(0); w++ {
```

Comment:
should we make this value configurable ? 🤔

Reply:
Ideally no :) I was thinking of moving it as a parameter in the caller if we want to, but because `pollProjects` is the main logic of our application it should execute parallel (note, not concurrent) operations for only `GOMAXPROCS` goroutines at once.
If other gorotuines are started by this they should have a `Context` attached, but we can review this part in the next PR :)

Reply:
sounds good! :)

Reply:
well given that the app is going to be run mostly in containers, maybe we could add a configuration parameters `workers` or something like that that defaults to `runtime.GOMAXPROCS(0)`

Reply:
indeed, this is what I was thinking of at first sight